alphalens/performance.py

```diff
@@ -247,7 +247,11 @@ def quantize_factor(factor, quantiles=5, by_group=False):
     """
 
     def quantile_calc(x, quantiles):
-        return pd.qcut(x, quantiles, labels=False) + 1
+        try:
+            return pd.qcut(x, quantiles, labels=False) + 1
+        except ValueError:
+            # This is supposed to fix "ValueError: Bin edges must be unique"
+            return pd.qcut(x.rank(method='first'), quantiles, labels=False) + 1
```

Comment:
Quoting [here](https://github.com/pandas-dev/pandas/issues/7751):

"Suppose we have a list with too many duplicates, say we want to split [1,2,3,3,3,3,3,3,4,5,6,7] into quartiles. Right now qcut fails, because the second-lowest quartile consists entirely of '3's, duplicating the bin edges. But there is a natural way to assign the quartiles if we allow duplicate values to be split among different bins: [1,2,3], [3,3,3], [3,3,4], [5,6,7]."

The same is true for this set of values  [1,1,1,1,1,2,2,3,3]. pandas.qcut fails with any number of quantiles in this scenario (see log below).

So, the point is that there might be cases where the bin edges fall in the middle of identical values and pandas.qcut will fail there. Ranking them will fix it, but we have that same values will go in different quantiles.

```
>>> pd.__version__
'0.17.1'
>>> s = pd.Series([1,1,1,1,1,2,2,3,3])
>>> pd.qcut(s,2, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([1, 1, 3])
>>> pd.qcut(s,3, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([ 1.,  1.,  2.,  3.])
>>> pd.qcut(s,4, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([1, 1, 1, 2, 3])
>>> pd.qcut(s,5, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([ 1. ,  1. ,  1. ,  1.8,  2.4,  3. ])
>>> pd.qcut(s,6, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([ 1.        ,  1.        ,  1.        ,  1.        ,  2.        ,
        2.66666667,  3.        ])
>>> pd.qcut(s,7, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([ 1.        ,  1.        ,  1.        ,  1.        ,  1.57142857,
        2.        ,  2.85714286,  3.        ])
>>> pd.qcut(s,8, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([1, 1, 1, 1, 1, 2, 2, 3, 3])
>>> len(s)
9
>>> pd.qcut(s,9, labels=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 169, in qcut
    include_lowest=True)
  File "/usr/lib/python2.7/dist-packages/pandas/tools/tile.py", line 189, in _bins_to_cuts
    raise ValueError('Bin edges must be unique: %s' % repr(bins))
ValueError: Bin edges must be unique: array([ 1.        ,  1.        ,  1.        ,  1.        ,  1.        ,
        1.44444444,  2.        ,  2.22222222,  3.        ,  3.        ])

```